contracts/protocol/tokenization/libraries/ApeStakingLogic.sol

```diff
@@ -373,95 +200,175 @@ library ApeStakingLogic {
+        {
+            //2.1 unstake Main pool position
             (uint256 stakedAmount, ) = _apeCoinStaking.nftPosition(
                 poolId,
-                tokenIds[index]
+                tokenId
             );
-
             if (stakedAmount > 0) {
-                nfts[counter] = ApeCoinStaking.SingleNft({
-                    tokenId: tokenIds[index],
-                    amount: stakedAmount
-                });
-                counter++;
+                ApeCoinStaking.SingleNft[]
+                    memory nfts = new ApeCoinStaking.SingleNft[](1);
+                nfts[0].tokenId = tokenId;
+                nfts[0].amount = stakedAmount;
+                if (poolId == BAYC_POOL_ID) {
+                    _apeCoinStaking.withdrawBAYC(nfts, address(this));
+                } else {
+                    _apeCoinStaking.withdrawMAYC(nfts, address(this));
+                }
             }
-
+            //2.2 unstake bakc pool position
             (uint256 bakcTokenId, bool isPaired) = _apeCoinStaking.mainToBakc(
                 poolId,
-                tokenIds[index]
+                tokenId
             );
             if (isPaired) {
-                _bakcPairs[index] = ApeCoinStaking.PairNftWithAmount({
-                    mainTokenId: tokenIds[index],
-                    bakcTokenId: bakcTokenId,
-                    amount: 0
-                });
-                pairdBakcCounter++;
+                (stakedAmount, ) = _apeCoinStaking.nftPosition(
+                    BAKC_POOL_ID,
+                    bakcTokenId
+                );
+                if (stakedAmount > 0) {
+                    ApeCoinStaking.PairNftWithAmount[]
+                        memory _nftPairs = new ApeCoinStaking.PairNftWithAmount[](
+                            1
+                        );
+                    _nftPairs[0].mainTokenId = tokenId;
+                    _nftPairs[0].bakcTokenId = bakcTokenId;
+                    _nftPairs[0].amount = stakedAmount;
+                    ApeCoinStaking.PairNftWithAmount[]
+                        memory _otherPairs = new ApeCoinStaking.PairNftWithAmount[](
+                            0
+                        );
+
+                    if (poolId == BAYC_POOL_ID) {
+                        _apeCoinStaking.withdrawBAKC(_nftPairs, _otherPairs);
+                    } else {
+                        _apeCoinStaking.withdrawBAKC(_otherPairs, _nftPairs);
+                    }
+                    IERC721(_apeCoinStaking.nftContracts(BAKC_POOL_ID))
+                        .transferFrom(
+                            address(this),
+                            positionOwner,
+                            bakcTokenId
+                        );
+                }
             }
         }
 
-        if (counter > 0) {
-            assembly {
-                mstore(nfts, counter)
+        IERC20 _apeCoin = _apeCoinStaking.apeCoin();
+        uint256 apeBalance = _apeCoin.balanceOf(address(this));
```

Comment:
should calculate the  apeBalance  as the (balance after - balance before) to be safe?